src/main/java/com/o19s/es/ltr/query/DerivedExpressionQuery.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright [2017] Wikimedia Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.o19s.es.ltr.query;
+
+import com.o19s.es.ltr.feature.FeatureSet;
+import com.o19s.es.ltr.ranker.LtrRanker;
+import org.apache.lucene.expressions.Bindings;
+import org.apache.lucene.expressions.Expression;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.DoubleValues;
+import org.apache.lucene.search.DoubleValuesSource;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+
+public class DerivedExpressionQuery extends Query {
+    private final FeatureSet features;
+    private final Expression expression;
+
+    public DerivedExpressionQuery(FeatureSet features, Expression expr) {
+        this.features = features;
+        this.expression = expr;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
+        if (!needsScores) {
+            // If scores are not needed simply return a constant score on all docs
+            return new ConstantScoreWeight(this) {
+                @Override
+                public Scorer scorer(LeafReaderContext context) throws IOException {
+                    return new ConstantScoreScorer(this, score(), DocIdSetIterator.all(context.reader().maxDoc()));
+                }
+            };
+        }
+
+        return new FVWeight(this);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (!sameClassAs(obj)) {
+            return false;
+        }
+        DerivedExpressionQuery that = (DerivedExpressionQuery) obj;
+        return Objects.deepEquals(expression, that.expression)
+                && Objects.deepEquals(features, that.features);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(expression, features);
```

Comment:
right I missed that one, not a big deal I think because hashCode&equals should never be called at runtime (they're meant for caching but we never cache those)
Maybe worth fixing on a followup PR, it's probably cleaner.